plutus-tx/src/PlutusTx/Numeric.hs

```diff
@@ -126,3 +146,33 @@ instance MultiplicativeSemigroup a => Semigroup (Product a) where
 instance MultiplicativeMonoid a => Monoid (Product a) where
     {-# INLINABLE mempty #-}
     mempty = Product one
+
+-- TODO. This is doing twice as much work as it needs to: the Plutus Core
+-- builtins 'divideInteger' and 'modInteger' are implemented as the Haskell
+-- 'div' and 'mod' operations, which in turn are impemented as fst . divMod and
+-- snd . divMod, so we're calling Haskell's 'divMod' twice and throwing away
+-- useful information.  We could fix his by exposing 'divMod' as a PLC builtin,
```

Comment:
I don't actually believe this is the case. `div` and `mod` typically have separate implementations which are put together to form `divMod`, and a quick look at the Core verifies that GHC is indeed specializing the implementation of the builtin functions to just call the underlying `div` or `mod`.

Reply:
This is a direct copy of the doc notes that already existed on this function.

Reply:
So it is. They're wrong!

Reply:
>  div and mod typically have separate implementations which are put together to form divMod

@michaelpj I believe that I wrote the note being discussed here, which says that `div` and `mod` are derived from `divMod`.  That's based on [this code](https://hackage.haskell.org/package/base-4.16.0.0/docs/src/GHC.Real.html#divMod) from `GHC,Real`.  Could you clarify what you're saying about the Core?  My point was that here we're implementing `divMod` as  `divMod x y = ( x `divideInteger` y, x `modInteger` y)` and the two functions being called on the right hand side will be calling Haskell's `divMod` function with the same arguments.

Reply:
The Haskell implementation of `divideInteger` uses the Haskell function `div`. GHC is smart enough to compile `div` into a division primitive, even though it is initially defined in terms of `divMod`. You can check the Core to see. So there is no extra overhead.

Reply:
> The Haskell implementation of `divideInteger` uses the Haskell function `div`.

Yes, but my point is that the Haskell function `div` is defined in terms of the Haskell function `divMod`. It's effectively `fst . divMod` and this is inherited from gmp: gmp only has one basic integer division function which calculates the quotient and remainder simultaneously, because there's no quick way to do one without the other.  The comment above refers to the _Plutus_ function `divMod`, which is defined to be `(x 'divideInteger' y, x 'modInteger' y)`: in Plutus Core that will currently compile to a Scott encoded pair which calls `divideInteger` and `modInteger`, and both of these will call _Haskell's_ `divMod`, doing twice as much work as necessary.  If we had `divMod` as a PLC builtin then we could avoid this, but that would have been difficult without built-in pairs.  The statement that

> div and mod typically have separate implementations which are put together to form divMod

is false for both Haskell and GMP: it's  the other way around.

Anyway, none of this has anything to do with this PR!  It's a comment that was written a long time ago and has become slightly outdated with the advent of built-in pairs.